### PR TITLE
`gcs-upload`: perform additional checks to see if data/file is valid for compression at write time

### PR DESCRIPTION
**Pull request by smg247**
This is a follow-up/replacement to https://github.com/kubernetes/test-infra/pull/32148. That functionality only worked for files rather than both files and data being uploaded to gcs. Instead of determining if the file/data should be compressed given only the dest, we utilize just-in-time logic to make the determination at write time. This solves the issue of making these determinations about data functional, and also improves performance for file determinations as we don't need an extra `stat` and `read` on each file.

Most of the changes here involve simplifying the unit tests to deal with the new architecture.

### CODE REVIEW

File: prow/pod-utils/gcs/upload.go [hunk 1]
```diff
@@ -266,28 +230,43 @@ type dataWriter interface {
...
+	compressFileType bool
+	opts             []pkgio.WriterOptions
+	writer           pkgio.Writer
+	closers          []pkgio.Closer
 }
 
 func (w *openerObjectWriter) Write(p []byte) (n int, err error) {
 	if w.writer == nil {
+		largerThanOneKB := len(p) > 1024
+		alreadyCompressed := http.DetectContentType(p) == "application/x-gzip"
```
Comment (stevekuznetsov) [inline prow/pod-utils/gcs/upload.go:249]: nit: this reads 512 bytes, when multiplies by 1k files you start to see an impact. I appreciate the named booleans but in-lining this into the conditional would defer the write until it was actually necessary.

**Review by stevekuznetsov — COMMENTED** [review]
/lgtm
/approve
/hold

Just the one nit, up to you.

**Comment by smg247** [comment]
@stevekuznetsov I fixed the nit (and a funny linter error) can you retag?

**Comment by stevekuznetsov** [comment]
/lgtm

@smg247 sorry! I was waiting for tests to pass, then got sidetracked.

**Pull request merged**